sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/GoneAndRetryWithRetryPolicy.java

```diff
@@ -16,189 +15,324 @@
-    private volatile int currentBackoffSeconds = GoneAndRetryWithRetryPolicy.INITIAL_BACKOFF_TIME;
-    private volatile RetryWithException lastRetryWithException;
+    private final GoneRetryPolicy goneRetryPolicy;
+    private final RetryWithRetryPolicy retryWithRetryPolicy;
     private final StopWatch durationTimer = new StopWatch();
-    private final int waitTimeInSeconds;
-    //TODO once this is moved to IRetryPolicy, remove from here.
-    public final static Quadruple<Boolean, Boolean, Duration, Integer> INITIAL_ARGUMENT_VALUE_POLICY_ARG = Quadruple.with(false, false,
-            Duration.ofSeconds(60), 0);
 
     public GoneAndRetryWithRetryPolicy(RxDocumentServiceRequest request, Integer waitTimeInSeconds) {
-        this.request = request;
+        this.goneRetryPolicy = new GoneRetryPolicy(request, waitTimeInSeconds, durationTimer);
+        this.retryWithRetryPolicy = new RetryWithRetryPolicy(request, waitTimeInSeconds, this.durationTimer);
         startStopWatch(this.durationTimer);
-        if (waitTimeInSeconds != null) {
-            this.waitTimeInSeconds = waitTimeInSeconds;
-        } else {
-            this.waitTimeInSeconds = DEFAULT_WAIT_TIME_IN_SECONDS;
-        }
     }
 
     @Override
     public Mono<ShouldRetryResult> shouldRetry(Exception exception) {
-        CosmosException exceptionToThrow = null;
-        Duration backoffTime = Duration.ofSeconds(0);
-        Duration timeout = Duration.ofSeconds(0);
-        boolean forceRefreshAddressCache = false;
-        if (!(exception instanceof GoneException) &&
-            !(exception instanceof RetryWithException) &&
-            !(exception instanceof PartitionIsMigratingException) &&
-            !(exception instanceof InvalidPartitionException &&
-                (this.request.getPartitionKeyRangeIdentity() == null ||
-                this.request.getPartitionKeyRangeIdentity().getCollectionRid() == null)) &&
-            !(exception instanceof PartitionKeyRangeIsSplittingException)) {
-
-            logger.warn("Operation will NOT be retried. Current attempt {}, Exception: ", this.attemptCount,
-                exception);
-            stopStopWatch(this.durationTimer);
-            return Mono.just(ShouldRetryResult.noRetry());
-        } else if (exception instanceof GoneException &&
-            !request.isReadOnly() &&
-            BridgeInternal.hasSendingRequestStarted((CosmosException)exception)) {
+
+        return this.retryWithRetryPolicy.shouldRetry(exception)
+                                        .flatMap((retryWithResult) -> {
+
+            if (retryWithResult.shouldRetry) {
+                return Mono.just(retryWithResult);
+            }
+
+            return this.goneRetryPolicy.shouldRetry(exception)
+                .flatMap((goneRetryResult) -> {
+                    if (!goneRetryResult.shouldRetry) {
+                        logger.warn("Operation will NOT be retried. Exception:",
+                            exception);
```

Comment:
this logs stacktrace with warning level.
we are logging stacktrace with warning level inside the inner policies too. this way we will get stacktrace dumped twice.
we should log stacktrace with warn only at one place. I wonder if we can remove this one and fully rely on the warn from the inner retryPolicy log?

Reply:
ACK - makes sense - will make it a debug level log here